Sources/RabbitMq/Connection.swift

```diff
@@ -1,46 +1,149 @@
+    }
+
+    // Method to use to connect without monitoring, will be called when using reuseChannel()
+    public func connect() async throws {
+        if !isConnected && !connecting {
+            connecting = true
+            defer { connecting = false }
+
+            logger.info("Connecting to broker at \(url)")
+            self.connection = try await AMQPConnection.connect(use: self.eventLoop, from: self.config)
+            logger.info("Connected to broker at \(url)")
         }
-        return false
     }
 
-    public func reuseChannel() async throws -> AMQPChannel {
-        guard let channel = self.channel, channel.isOpen else {
-            if self.connection == nil || self.connection!.isConnected {
-                self.connection = try await AMQPConnection.connect(
-                    use: self.eventLoop, from: self.config)
+    private func gracefulCancellableDelay(timeout: Duration) async throws {
+        for await _ in AsyncTimerSequence(interval: timeout, clock: .continuous).cancelOnGracefulShutdown() {
+            break
+        }
+    }
+
+    private func monitorConnection(reconnectionInterval: Duration) async throws {
+        while !Task.isCancelled && !Task.isShuttingDownGracefully {
+            // Ignore if connected
+            if isConnected {
+                try await Task.sleep(for: PollingConnectionSleepInterval)
+                continue
             }
 
+            // Connect or reconnect after interval
+            do {
+                try await self.connect()
+                continue
+            } catch {
+                logger.error("Unable to connect to broker at \(self.url): \(error)")
+                try await gracefulCancellableDelay(timeout: reconnectionInterval)
+            }
+        }
+    }
+
+    public func run(reconnectionInterval: Duration?) async throws {
+        try await withThrowingDiscardingTaskGroup { group in
+            if let interval = reconnectionInterval {
+                group.addTask {
+                    // Monitor connection task
+                    try await self.monitorConnection(reconnectionInterval: interval)
+
+                    // Stop receiving new consumers
+                    self.newConsumers.finish()
+
+                    // Close connection at the end
+                    try? await self.close()
+                }
+            }
+
+            for try await consumer in newConsumers {
+                group.addTask { try await consumer.run() }
+            }
+
+            // Cancel all tasks on exit
+            group.cancelAll()
+        }
+    }
+
+    public func getChannel() async throws -> AMQPChannel? {
+        // Not connected
+        guard isConnected else {
+            return nil
+        }
+
+        // We're connected, let's reuse the channel
+        guard let channel = self.channel, channel.isOpen else {
+            // Then open a channel
             self.channel = try await connection!.openChannel()
```

Comment:
This has the same race condition we've seen before.

Reply:
Well that's fun. Maybe I can implement something similar on the actor to remove the race condition.